pydantic/_internal/_generate_schema.py

```diff
@@ -766,3 +771,137 @@ def get_model_self_schema(cls: type[BaseModel]) -> core_schema.ModelSchema:
         core_schema.definition_reference_schema(model_ref),
         metadata=build_metadata_dict(js_metadata=model_js_metadata),
     )
+
+
+def apply_discriminator(schema: core_schema.CoreSchema, discriminator: str) -> core_schema.CoreSchema:
+    # Eventually: should add support for other discriminator types, and explicitly specified choices
+
+    # If the field was wrapped with a nullable schema, we preserve that here and re-wrap the final result at the end.
+    nullable_schema: core_schema.NullableSchema | None = None
+    if schema['type'] == 'nullable':
+        nullable_schema = schema
+        schema = schema['schema']
+
+    if schema['type'] != 'union':
+        raise TypeError('`discriminator` can only be used with `Union` type with more than one variant')
+
+    union_choices = schema['choices']
+    if len(union_choices) < 2:
+        raise TypeError('`discriminator` can only be used with `Union` type with more than one variant')
+
+    # `aliases` is meant to behave like a set, but uses a dict to ensure order is preserved
+    aliases = {discriminator: None}
+    tagged_union_choices = _build_tagged_union_choices(union_choices, discriminator, aliases)
+
+    if len(aliases) > 1:
+        schema_discriminator: str | list[list[str | int]] = [[alias] for alias in aliases]
```

Comment:
Yes, I think this is described in pydantic_core somewhere (I wanted documentation because it confused the crap out of me when I first saw it), but I'll add a note here about what it means. The idea for discriminators in general:
* `str`: name of property to grab to obtain discriminator value
* `int`: index to grab (for arrays) to obtain discriminator value
* `list[str | int]`: obtain the discriminator value by iteratively/recursively applying the items of the list to the value obtained following the instructions in the first two bullets above
* `list[list[str | int]]`: follow the procedure described in the previous bullet for each item in the outer list until you get a valid discriminator value

So, if you need to check multiple fields to grab the discriminator value (as is the case if populate_by_name is True and there is an alias), the way to do this is to use the list-of-lists approach (representing alternative choices) where the inner lists represent accessing the specific property that you want to grab the discriminator from.

Does that make sense?

Reply:
Wow, I mean yeah that makes sense but I'm surprised we support something that complex. Do other libraries (zod, etc.) support this sort of recursive discriminant stuff? I thought most only supported the "give me the name of a property, it better be a scalar value" type thing.

Reply:
Most things that work with discriminators definitely work that way, I think Samuel implemented this in pursuit of generality. That said, I think this will allow people to create nice wrappers around certain "crappy" APIs that don't do "standard" things with discriminators.

Also note: the pydantic_core TaggedUnion thing actually even allows arbitrary python functions to be used to obtain the discriminator from a value. I haven't added any support for that now though.